pkg/restapi/v1/verifier/controller.go

```diff
@@ -634,57 +640,93 @@ func validateIDToken(rawJwt string, verifier jose.SignatureVerifier) (*IDTokenCl
 }
-func validateVPToken(rawJwt string, verifier jose.SignatureVerifier) (*VPTokenClaims, string, error) {
-	_, signer, rawClaims, err := verifyTokenSignature(
-		rawJwt,
-		jwt.WithSignatureVerifier(verifier),
-		jwt.WithIgnoreClaimsMapDecoding(true),
-	)
+
+func (c *Controller) validateRawVPToken(vpToken string) (*VPTokenClaims, error) {
+	if jwt.IsJWS(vpToken) {
+		return c.validateVPTokenJWT(vpToken)
+	}
+
+	return c.validateVPTokenLDP(vpToken)
+}
+
+func (c *Controller) validateVPTokenJWT(vpToken string) (*VPTokenClaims, error) {
+	jsonWebToken, rawClaims, err := jwt.Parse(vpToken,
+		jwt.WithSignatureVerifier(c.jwtVerifier),
+		jwt.WithIgnoreClaimsMapDecoding(true))
 	if err != nil {
-		return nil, "", resterr.NewValidationError(resterr.InvalidValue, "vp_token", err)
+		return nil, resterr.NewValidationError(resterr.InvalidValue, "vp_token", err)
 	}
 
 	var fastParser fastjson.Parser
 	v, err := fastParser.ParseBytes(rawClaims)
 	if err != nil {
-		return nil, "", fmt.Errorf("decode claims: %w", err)
-	}
-
-	vpTokenClaims := &VPTokenClaims{
-		Nonce: string(v.GetStringBytes("nonce")),
-		Exp:   v.GetInt64("exp"),
+		return nil, fmt.Errorf("decode claims: %w", err)
 	}
 
-	vpData := v.Get("vp")
-	vpDataRaw, err := vpData.Object()
-	if vpData.Type() != fastjson.TypeNull && err != nil {
-		return nil, "", fmt.Errorf("decode claims2: %w", err)
-	}
-	if vpDataRaw != nil {
-		vpTokenClaims.VP = []byte(vpDataRaw.String())
-	}
-	if vpTokenClaims.Exp < time.Now().Unix() {
-		return nil, "", resterr.NewValidationError(resterr.InvalidValue, "vp_token.exp", fmt.Errorf(
+	exp := v.GetInt64("exp")
+	if exp < time.Now().Unix() {
+		return nil, resterr.NewValidationError(resterr.InvalidValue, "vp_token.exp", fmt.Errorf(
 			"token expired"))
 	}
 
-	if vpTokenClaims.VP == nil {
-		return nil, "", resterr.NewValidationError(resterr.InvalidValue, "vp_token.vp", fmt.Errorf(
-			"$vp is missed"))
+	presentation, err := verifiable.ParsePresentation([]byte(vpToken),
+		verifiable.WithPresJSONLDDocumentLoader(c.documentLoader),
+		verifiable.WithPresPublicKeyFetcher(
+			verifiable.NewVDRKeyResolver(c.vdr).PublicKeyFetcher(),
+		))
+	if err != nil {
+		return nil, resterr.NewValidationError(resterr.InvalidValue, "vp_token.vp", err)
 	}
 
-	return vpTokenClaims, signer, nil
+	// Do not need to check since is has passed jwt.Parse().
+	kid, _ := jsonWebToken.Headers.KeyID()
+
+	return &VPTokenClaims{
+		Nonce:         string(v.GetStringBytes("nonce")),
+		Aud:           string(v.GetStringBytes("aud")),
+		SignerDIDID:   strings.Split(kid, "#")[0],
+		VpTokenFormat: vcsverifiable.Jwt,
+		VP:            presentation,
+	}, nil
 }
 
-func verifyTokenSignature(rawJwt string, parseOps ...jwt.ParseOpt) (*jwt.JSONWebToken, string, []byte, error) {
-	jsonWebToken, rawClaims, err := jwt.Parse(rawJwt, parseOps...)
+func (c *Controller) validateVPTokenLDP(vpToken string) (*VPTokenClaims, error) {
+	presentation, err := verifiable.ParsePresentation([]byte(vpToken),
```

Comment:
Unfortunately no. I need to have parsed presentation for the future token processing.
It shouldn't be any performance issues with this version since the amount of calls to `verifiable.ParsePresentation()` stays the same - I have just moved them separately for JWT and JSON-LD cases.